File: examples/mandelbrot-demo/python/mandelbrot_pure.py
"""Pure Python Mandelbrot implementation (no C++ bindings)"""

import math

def escape_time(real, imag, max_iter):
    """Compute escape time for a single point with smooth coloring"""
    zr, zi = 0.0, 0.0
    iter = 0

    while zr * zr + zi * zi <= 4.0 and iter < max_iter:
        temp = zr * zr - zi * zi + real
        zi = 2.0 * zr * zi + imag
        zr = temp
        iter += 1

    if iter == max_iter:
        return float(max_iter)

    # Smooth coloring
    log_zn = math.log(zr * zr + zi * zi) / 2.0
    nu = math.log(log_zn / math.log(2.0)) / math.log(2.0)
    return float(iter) + 1.0 - nu

def iter_to_rgb(iter_val, max_iter):
    """Convert iteration count to RGB color"""
    if iter_val >= max_iter:
        return (0, 0, 0)

    t = iter_val / max_iter
    hue = (t * 5.0 + 0.6) % 1.0
    sat = 0.85
    val = t / 0.02 if t < 0.02 else 1.0

    # HSV to RGB
    hi = int(hue * 6.0) % 6
    f = hue * 6.0 - hi
    p = val * (1.0 - sat)
    q = val * (1.0 - f * sat)
    tv = val * (1.0 - (1.0 - f) * sat)

    if hi == 0:
        rd, gd, bd = val, tv, p
    elif hi == 1:
        rd, gd, bd = q, val, p
    elif hi == 2:
        rd, gd, bd = p, val, tv
    elif hi == 3:
        rd, gd, bd = p, q, val
    elif hi == 4:
        rd, gd, bd = tv, p, val
    else:
        rd, gd, bd = val, p, q

    return (int(rd * 255), int(gd * 255), int(bd * 255))

def render_mandelbrot(width, height, max_iter, x_min, x_max, y_min, y_max):
    """Render Mandelbrot to RGB pixel list - THE HOT LOOP"""
    pixels = []

    x_scale = (x_max - x_min) / width
    y_scale = (y_max - y_min) / height

    for py in range(height):
        imag = y_min + py * y_scale
        for px in range(width):
            real = x_min + px * x_scale
            iter_val = escape_time(real, imag, max_iter)
            r, g, b = iter_to_rgb(iter_val, max_iter)
            pixels.extend([r, g, b])

    return pixels

def to_ppm(width, height, pixels):
    """Convert pixel data to PPM string"""
    ppm = f"P3\n{width} {height}\n255\n"
    for i in range(0, len(pixels), 3):
        ppm += f"{pixels[i]} {pixels[i+1]} {pixels[i+2]}"
        ppm += "\n" if ((i // 3) % width == width - 1) else " "
    return ppm

# For benchmarking
if __name__ == "__main__":
    import time

    width, height = 800, 600
    max_iter = 256
    x_min, x_max = -2.5, 1.0
    y_min, y_max = -1.25, 1.25

    print(f"Pure Python Mandelbrot ({width}x{height}, {max_iter} iterations)")

    start = time.perf_counter()
    pixels = render_mandelbrot(width, height, max_iter, x_min, x_max, y_min, y_max)
    elapsed = time.perf_counter() - start

    print(f"Render time: {elapsed:.3f} seconds")
    print(f"Pixels: {len(pixels) // 3}")

File: examples/mandelbrot-demo/lua/mandelbrot_pure.lua
-- Pure Lua Mandelbrot implementation (no C++ bindings)

local function escape_time(real, imag, max_iter)
    -- Compute escape time for a single point with smooth coloring
    local zr, zi = 0.0, 0.0
    local iter = 0

    while zr * zr + zi * zi <= 4.0 and iter < max_iter do
        local temp = zr * zr - zi * zi + real
        zi = 2.0 * zr * zi + imag
        zr = temp
        iter = iter + 1
    end

    if iter == max_iter then
        return max_iter
    end

    -- Smooth coloring
    local log_zn = math.log(zr * zr + zi * zi) / 2.0
    local nu = math.log(log_zn / math.log(2.0)) / math.log(2.0)
    return iter + 1.0 - nu
end

local function iter_to_rgb(iter_val, max_iter)
    -- Convert iteration count to RGB color
    if iter_val >= max_iter then
        return 0, 0, 0
    end

    local t = iter_val / max_iter
    local hue = (t * 5.0 + 0.6) % 1.0
    local sat = 0.85
    local val = t < 0.02 and t / 0.02 or 1.0

    -- HSV to RGB
    local hi = math.floor(hue * 6.0) % 6
    local f = hue * 6.0 - hi
    local p = val * (1.0 - sat)
    local q = val * (1.0 - f * sat)
    local tv = val * (1.0 - (1.0 - f) * sat)

    local rd, gd, bd
    if hi == 0 then
        rd, gd, bd = val, tv, p
    elseif hi == 1 then
        rd, gd, bd = q, val, p
    elseif hi == 2 then
        rd, gd, bd = p, val, tv
    elseif hi == 3 then
        rd, gd, bd = p, q, val
    elseif hi == 4 then
        rd, gd, bd = tv, p, val
    else
        rd, gd, bd = val, p, q
    end

    return math.floor(rd * 255), math.floor(gd * 255), math.floor(bd * 255)
end

local function render_mandelbrot(width, height, max_iter, x_min, x_max, y_min, y_max)
    -- Render Mandelbrot to RGB pixel table - THE HOT LOOP
    local pixels = {}

    local x_scale = (x_max - x_min) / width
    local y_scale = (y_max - y_min) / height

    for py = 0, height - 1 do
        local imag = y_min + py * y_scale
        for px = 0, width - 1 do
            local real = x_min + px * x_scale
            local iter_val = escape_time(real, imag, max_iter)
            local r, g, b = iter_to_rgb(iter_val, max_iter)
            pixels[#pixels + 1] = r
            pixels[#pixels + 1] = g
            pixels[#pixels + 1] = b
        end
    end

    return pixels
end

local function to_ppm(width, height, pixels)
    -- Convert pixel data to PPM string
    local parts = {string.format("P3\n%d %d\n255\n", width, height)}

    for i = 1, #pixels, 3 do
        local idx = (i - 1) / 3
        parts[#parts + 1] = string.format("%d %d %d", pixels[i], pixels[i+1], pixels[i+2])
        if idx % width == width - 1 then
            parts[#parts + 1] = "\n"
        else
            parts[#parts + 1] = " "
        end
    end

    return table.concat(parts)
end

-- Export for use as module
local M = {
    escape_time = escape_time,
    iter_to_rgb = iter_to_rgb,
    render_mandelbrot = render_mandelbrot,
    to_ppm = to_ppm
}

-- For benchmarking when run directly
if arg and arg[0] and arg[0]:match("mandelbrot_pure") then
    local width, height = 800, 600
    local max_iter = 256
    local x_min, x_max = -2.5, 1.0
    local y_min, y_max = -1.25, 1.25

    print(string.format("Pure Lua Mandelbrot (%dx%d, %d iterations)", width, height, max_iter))

    local start = os.clock()
    local pixels = render_mandelbrot(width, height, max_iter, x_min, x_max, y_min, y_max)
    local elapsed = os.clock() - start

    print(string.format("Render time: %.3f seconds", elapsed))
    print(string.format("Pixels: %d", #pixels / 3))
end

return M
